File: testdata/fixtures/bench/javascript/safe/parameterized_sql.ts
import { Request, Response } from 'express';
import { Pool } from 'pg';
import knex from 'knex';

const pool = new Pool();
const db = knex({ client: 'pg' });

// SAFE: Parameterized query with pg pool - uses $1 placeholder
export async function getUserById(req: Request, res: Response): Promise<void> {
  const userId = req.params.id;
  const result = await pool.query(
    'SELECT id, name, email FROM users WHERE id = $1 AND deleted_at IS NULL',
    [userId]
  );
  res.json(result.rows);
}

// SAFE: Knex query builder - automatically parameterized
export async function searchProducts(req: Request, res: Response): Promise<void> {
  const term = req.query.q as string;
  const category = req.query.category as string;

  const products = await db('products')
    .select('id', 'name', 'price', 'description')
    .where('name', 'ilike', '%' + term + '%')
    .andWhere('category', category)
    .andWhere('active', true)
    .orderBy('name')
    .limit(50);

  res.json({ data: products });
}

// SAFE: Multi-parameter query with pg - all placeholders
export async function getOrderHistory(req: Request, res: Response): Promise<void> {
  const userId = req.query.userId as string;
  const status = req.query.status as string;
  const limit = parseInt(req.query.limit as string, 10) || 20;

  const { rows } = await pool.query(
    'SELECT o.id, o.total, o.status, o.created_at FROM orders o WHERE o.user_id = $1 AND o.status = $2 ORDER BY o.created_at DESC LIMIT $3',
    [userId, status, limit]
  );

  res.json(rows);
}

// SAFE: INSERT with parameterized query
export async function createUser(req: Request, res: Response): Promise<void> {
  const { name, email } = req.body;
  const result = await pool.query(
    'INSERT INTO users (name, email, created_at) VALUES ($1, $2, NOW()) RETURNING id',
    [name, email]
  );
  res.status(201).json({ id: result.rows[0].id });
}

// SAFE: UPDATE with parameterized query
export async function updateProfile(req: Request, res: Response): Promise<void> {
  const userId = req.params.id;
  const { displayName, bio } = req.body;
  await pool.query(
    'UPDATE profiles SET display_name = $1, bio = $2, updated_at = NOW() WHERE user_id = $3',
    [displayName, bio, userId]
  );
  res.json({ success: true });
}

// SAFE: DELETE with parameterized query
export async function deleteComment(req: Request, res: Response): Promise<void> {
  const commentId = req.params.commentId;
  const userId = req.query.userId as string;
  await pool.query(
    'DELETE FROM comments WHERE id = $1 AND author_id = $2',
    [commentId, userId]
  );
  res.status(204).send();
}

File: batou-rules/testdata/fixtures/bench/go/safe/parameterized_sql.go
package safe

import (
	"database/sql"
	"encoding/json"
	"net/http"
	"strconv"
)

// SAFE: Parameterized query with database/sql placeholder $1
func GetUserByID(db *sql.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		userID := r.URL.Query().Get("id")
		row := db.QueryRow(
			"SELECT id, name, email FROM users WHERE id = $1 AND active = true",
			userID,
		)
		var id int
		var name, email string
		if err := row.Scan(&id, &name, &email); err != nil {
			http.Error(w, "not found", http.StatusNotFound)
			return
		}
		json.NewEncoder(w).Encode(map[string]interface{}{
			"id": id, "name": name, "email": email,
		})
	}
}

// SAFE: Parameterized INSERT
func CreateUser(db *sql.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		name := r.FormValue("name")
		email := r.FormValue("email")
		var id int
		err := db.QueryRow(
			"INSERT INTO users (name, email) VALUES ($1, $2) RETURNING id",
			name, email,
		).Scan(&id)
		if err != nil {
			http.Error(w, "insert failed", http.StatusInternalServerError)
			return
		}
		w.WriteHeader(http.StatusCreated)
		json.NewEncoder(w).Encode(map[string]int{"id": id})
	}
}

// SAFE: Parameterized search with LIMIT
func SearchProducts(db *sql.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		term := r.URL.Query().Get("q")
		limitStr := r.URL.Query().Get("limit")
		limit, err := strconv.Atoi(limitStr)
		if err != nil || limit < 1 || limit > 100 {
			limit = 20
		}
		rows, err := db.Query(
			"SELECT id, name, price FROM products WHERE name ILIKE $1 ORDER BY name LIMIT $2",
			"%"+term+"%", limit,
		)
		if err != nil {
			http.Error(w, "query failed", http.StatusInternalServerError)
			return
		}
		defer rows.Close()

		type product struct {
			ID    int     `json:"id"`
			Name  string  `json:"name"`
			Price float64 `json:"price"`
		}
		var results []product
		for rows.Next() {
			var p product
			rows.Scan(&p.ID, &p.Name, &p.Price)
			results = append(results, p)
		}
		json.NewEncoder(w).Encode(results)
	}
}
